core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -1674,6 +1678,18 @@ class LogCleanerTest {
   private def recoverAndCheck(config: LogConfig, expectedKeys: Iterable[Long]): Log = {
     LogTest.recoverAndCheck(dir, config, expectedKeys, new BrokerTopicStats(), time, time.scheduler)
   }
+
+  /**
+   * We need to run a two pass clean to perform the following steps to stimulate a proper clean:
+   *  1. On the first run, set the delete horizon stored in the first timestamp of the batches in the log.
```

Comment:
Perhaps it's clearer with "On the first run, set the delete horizon in the batches with tombstone or markers with empty txn records."